File: src/praisonai-agents/praisonaiagents/frameworks/protocols.py
"""Protocol for optional third-party agent framework adapters."""

from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional, Protocol, runtime_checkable


@runtime_checkable
class FrameworkAdapterProtocol(Protocol):
    """Contract for YAML-driven multi-framework execution backends."""

    name: str
    install_hint: str
    requires_tools_extra: bool
    is_router: bool

    def is_available(self) -> bool:
        """Return True when this framework's optional dependencies are installed."""
        ...

    def resolve(
        self, *, config: Optional[Dict[str, Any]] = None
    ) -> "FrameworkAdapterProtocol":
        """Pick a concrete adapter variant (e.g. autogen v0.2 vs v0.4)."""
        ...

    def setup(self, *, framework_tag: str) -> None:
        """Framework-specific pre-run hooks (observability, SDK init, etc.)."""
        ...

    def run(
        self,
        config: Dict[str, Any],
        llm_config: List[Dict],
        topic: str,
        *,
        tools_dict: Optional[Dict[str, Any]] = None,
        agent_callback: Optional[Callable] = None,
        task_callback: Optional[Callable] = None,
        cli_config: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Execute the framework with shared agents.yaml configuration."""
        ...

    async def arun(
        self,
        config: Dict[str, Any],
        llm_config: List[Dict],
        topic: str,
        *,
        tools_dict: Optional[Dict[str, Any]] = None,
        agent_callback: Optional[Callable] = None,
        task_callback: Optional[Callable] = None,
        cli_config: Optional[Dict[str, Any]] = None,
    ) -> str:
        """Async execution; default implementations may offload sync run()."""
        ...

    def cleanup(self) -> None:
        """Release resources after execution."""
        ...

    def resolve_alias(self) -> str:
        """Return the concrete adapter name to dispatch to."""
        ...

File: src/praisonai-ts/src/frameworks/protocols.ts
/**
 * Protocol for optional third-party agent framework adapters.
 *
 * Python parity: praisonaiagents/frameworks/protocols.py
 *
 * A framework adapter is a YAML-driven multi-framework execution backend
 * (autogen, crewai, ...). The contract is intentionally light so adapters can
 * be typed against it without pulling in any framework SDK.
 */

/** Keyword arguments of {@link FrameworkAdapterProtocol.resolve}. */
export interface FrameworkResolveOptions {
  config?: Record<string, unknown> | null;
}

/** Keyword arguments of {@link FrameworkAdapterProtocol.setup}. */
export interface FrameworkSetupOptions {
  frameworkTag: string;
}

/** Keyword arguments of {@link FrameworkAdapterProtocol.run} / `arun`. */
export interface FrameworkRunOptions {
  toolsDict?: Record<string, unknown> | null;
  agentCallback?: ((...args: unknown[]) => unknown) | null;
  taskCallback?: ((...args: unknown[]) => unknown) | null;
  cliConfig?: Record<string, unknown> | null;
}

/** Shared `llm_config` entries handed to a framework run. */
export type FrameworkLLMConfig = Array<Record<string, unknown>>;

/**
 * Contract for YAML-driven multi-framework execution backends.
 * Python parity: `FrameworkAdapterProtocol` (runtime_checkable Protocol).
 */
export interface FrameworkAdapterProtocol {
  name: string;
  installHint: string;
  requiresToolsExtra: boolean;
  isRouter: boolean;

  /** True when this framework's optional dependencies are installed. */
  isAvailable(): boolean;

  /** Pick a concrete adapter variant (e.g. autogen v0.2 vs v0.4). */
  resolve(options?: FrameworkResolveOptions): FrameworkAdapterProtocol;

  /** Framework-specific pre-run hooks (observability, SDK init, etc.). */
  setup(options: FrameworkSetupOptions): void;

  /** Execute the framework with shared agents.yaml configuration. */
  run(
    config: Record<string, unknown>,
    llmConfig: FrameworkLLMConfig,
    topic: string,
    options?: FrameworkRunOptions,
  ): string | Promise<string>;

  /** Async execution; default implementations may offload sync run(). */
  arun(
    config: Record<string, unknown>,
    llmConfig: FrameworkLLMConfig,
    topic: string,
    options?: FrameworkRunOptions,
  ): Promise<string>;

  /** Release resources after execution. */
  cleanup(): void;

  /** Return the concrete adapter name to dispatch to. */
  resolveAlias(): string;
}
